File: packages/core/src/determinism/discovery.ts
import { computeArtifactDigest } from "../identity/artifact-digest.js";
import { createOrderingPolicy } from "./context.js";

/**
 * Deterministic repository discovery and ownership resolution (Stage 3 —
 * Deterministic Execution & Reproducibility, ADR 0029).
 *
 * Baseline LLM repository discovery is repeatable: search/discovery →
 * classify relevance → normalize paths → stable rank/order → apply bounds.
 * Same task + same repository state → same initial deterministic
 * discovery result; filesystem enumeration order never decides the
 * baseline selection. The model may request additional exploration, but
 * the baseline is host-owned.
 */

export type DiscoveryRelevance = "verified" | "candidate";

export interface DiscoveryCandidate {
  readonly path: string;
  readonly relevance: DiscoveryRelevance;
  /** Bounded classification evidence (e.g. "exact task target", "direct import"). */
  readonly evidence: readonly string[];
  /** Rank signal: lower is more relevant (host-derived, never model claim). */
  readonly rank: number;
}

export interface DiscoveryResult {
  readonly candidates: readonly DiscoveryCandidate[];
  /** Canonical digest over the ordered candidates. */
  readonly digest: string;
}

export interface DiscoveryInput {
  readonly unorderedCandidates: readonly {
    readonly path: string;
    readonly relevance: DiscoveryRelevance;
    readonly evidence: readonly string[];
  }[];
  readonly maxCandidates: number;
  /** Normalized workspace-relative target paths of the task. */
  readonly taskTargets: readonly string[];
}

/**
 * Deterministic baseline discovery: exact task targets rank first
 * (strong evidence class), then verified, then candidate; ties break by
 * canonical path; bounds apply last. Shuffled input order cannot change
 * the result.
 */
export function discoverRepository(input: DiscoveryInput): DiscoveryResult {
  const ordering = createOrderingPolicy();
  const ranked: DiscoveryCandidate[] = input.unorderedCandidates.map((candidate) => {
    const exactTarget = input.taskTargets.includes(candidate.path);
    const strongEvidence =
      candidate.relevance === "verified" ||
      candidate.evidence.some((item) =>
        [
          "exact task target",
          "direct import",
          "structural relationship",
          "explicit architecture ownership",
          "test mapping",
        ].includes(item),
      );
    const rank = exactTarget ? 0 : strongEvidence ? 1 : 2;
    return {
      path: candidate.path,
      relevance: candidate.relevance,
      evidence: [...candidate.evidence],
      rank,
    };
  });
  const ordered = ordering.stableSort(ranked, (candidate) => {
    const rankKey = String(candidate.rank).padStart(4, "0");
    return `${rankKey}:${candidate.path}`;
  });
  const bounded = ordered.slice(0, input.maxCandidates);
  const digest = computeArtifactDigest({
    artifactType: "DiscoveryResult",
    schemaVersion: 1,
    payload: {
      candidates: bounded.map((candidate) => ({
        path: candidate.path,
        relevance: candidate.relevance,
        rank: candidate.rank,
        evidence: candidate.evidence,
      })),
    },
  }).value;
  return { candidates: bounded, digest };
}

// ---------------------------------------------------------------------------
// Ownership resolution
// ---------------------------------------------------------------------------

/**
 * Canonical subsystem ownership metadata (architecture/navigation only —
 * never a service registry, never capability). Before an executor creates
 * an overlapping abstraction, context identifies the existing owner.
 */
export interface OwnershipEntry {
  /** Canonical responsibility identifier, e.g. "tool projection". */
  readonly responsibility: string;
  /** Canonical owner module/class, e.g. "ToolProjector". */
  readonly owner: string;
  /** Source path of the owner. */
  readonly path: string;
  /** Related responsibilities that overlap and must reuse the owner. */
  readonly overlapsWith: readonly string[];
}

export const OWNERSHIP_INDEX: readonly OwnershipEntry[] = [
  {
    responsibility: "tool projection",
    owner: "ToolProjector",
    path: "packages/core/src/projection/tool-projector.ts",
    overlapsWith: ["provider tool schema"],
  },
  {
    responsibility: "context projection",
    owner: "ContextProjector",
    path: "packages/core/src/projection/context-projector.ts",
    overlapsWith: ["provider context assembly"],
  },
  {
    responsibility: "planning depth",
    owner: "PlanningPolicy",
    path: "packages/core/src/planning/planning-policy.ts",
    overlapsWith: ["plan routing"],
  },
  {
    responsibility: "task state",
    owner: "TaskRuntime",
    path: "packages/core/src/tasks/task-runtime.ts",
    overlapsWith: ["workflow state"],
  },
  {
    responsibility: "acceptance evaluation",
    owner: "AcceptanceEvaluator",
    path: "packages/core/src/executor/",
    overlapsWith: ["completion gate"],
  },
  {
    responsibility: "evidence",
    owner: "EvidenceStore",
    path: "packages/core/src/tasks/task-runtime-evidence.ts",
    overlapsWith: ["evidence records"],
  },
  {
    responsibility: "approval",
    owner: "ApprovalSystem",
    path: "packages/core/src/security/approval.ts",
    overlapsWith: ["mutation authorization"],
  },
  {
    responsibility: "checkpoints",
    owner: "CheckpointStore",
    path: "packages/adapters/src/checkpoints/",
    overlapsWith: ["undo", "recovery"],
  },
  {
    responsibility: "workspace revisions",
    owner: "WorkspaceRevisionRegistry",
    path: "packages/core/src/workspace/workspace-revision.ts",
    overlapsWith: ["source identity"],
  },
  {
    responsibility: "canonical digests",
    owner: "ArtifactDigest",
    path: "packages/core/src/identity/artifact-digest.ts",
    overlapsWith: ["hashing"],
  },
  {
    responsibility: "documentation selection",
    owner: "DocumentationSelection",
    path: "packages/core/src/executor/documentation-context.ts",
    overlapsWith: ["guidance selection"],
  },
  {
    responsibility: "executor briefing",
    owner: "ExecutorBriefCompiler",
    path: "packages/core/src/executor/brief-compiler.ts",
    overlapsWith: ["task briefing"],
  },
  {
    responsibility: "validation plan",
    owner: "ValidationPlan",
    path: "packages/core/src/determinism/decisions.ts",
    overlapsWith: ["validation selection"],
  },
  {
    responsibility: "retry policy",
    owner: "RetryPolicy",
    path: "packages/core/src/determinism/decisions.ts",
    overlapsWith: ["repair loop"],
  },
  {
    responsibility: "impact analysis",
    owner: "ImpactAnalyzer",
    path: "packages/core/src/godot/impact/impact-analyzer.ts",
    overlapsWith: ["review context"],
  },
  {
    responsibility: "independent review",
    owner: "ChangeReviewer",
    path: "packages/core/src/godot/quality/quality-review.ts",
    overlapsWith: ["quality gate"],
  },
  {
    responsibility: "surface routing",
    owner: "DevelopmentSurfaceClassifier",
    path: "packages/core/src/godot/development/development-surface.ts",
    overlapsWith: ["workflow routing"],
  },
  {
    responsibility: "nondeterminism audit",
    owner: "NondeterminismAudit",
    path: "scripts/check-nondeterminism.mjs",
    overlapsWith: ["architecture checks"],
  },
];

/** Deterministic ownership resolution: same responsibility → same owner. */
export function resolveOwner(responsibility: string): OwnershipEntry | null {
  const normalized = responsibility.trim().toLowerCase();
  const exact = OWNERSHIP_INDEX.find((entry) => entry.responsibility.toLowerCase() === normalized);
  if (exact !== undefined) {
    return exact;
  }
  return (
    OWNERSHIP_INDEX.find((entry) =>
      entry.overlapsWith.some((overlap) => overlap.toLowerCase() === normalized),
    ) ?? null
  );
}

/** Bounded deterministic listing of all canonical owners (stable order). */
export function listOwnership(): readonly OwnershipEntry[] {
  return createOrderingPolicy().stableSort(OWNERSHIP_INDEX, (entry) => entry.responsibility);
}

File: crates/siralos-core/src/determinism/discovery.rs
//! Deterministic repository discovery and ownership resolution
//! (Stage 3 — Deterministic Execution & Reproducibility, ADR 0029;
//! R10a H2).
//!
//! Mirrors `packages/core/src/determinism/discovery.ts`. Baseline
//! discovery is repeatable: search → classify relevance → stable
//! rank/order → apply bounds. Same task + same repository state → same
//! result; filesystem enumeration order never decides the baseline.
//! The model may request additional exploration, but the baseline is
//! host-owned.

use super::ports::stable_sort_by_key;
use crate::identity::{
    ArtifactIdentityError, CanonicalValue, compute_artifact_digest,
};
use std::collections::BTreeMap;

/// Relevance class of one discovery candidate.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DiscoveryRelevance {
    /// Host-verified relevance.
    Verified,
    /// Candidate (unverified).
    Candidate,
}

impl DiscoveryRelevance {
    /// Canonical protocol string.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Verified => "verified",
            Self::Candidate => "candidate",
        }
    }

    /// Parse a protocol string; unknown values are rejected.
    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "verified" => Some(Self::Verified),
            "candidate" => Some(Self::Candidate),
            _ => None,
        }
    }
}

/// One discovery candidate before ranking.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DiscoveryCandidateInput {
    /// Workspace-relative path.
    pub path: String,
    /// Declared relevance.
    pub relevance: DiscoveryRelevance,
    /// Bounded classification evidence.
    pub evidence: Vec<String>,
}

/// One ranked candidate in a discovery result.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DiscoveryCandidate {
    /// Workspace-relative path.
    pub path: String,
    /// Relevance class.
    pub relevance: DiscoveryRelevance,
    /// Bounded classification evidence.
    pub evidence: Vec<String>,
    /// Rank signal: lower is more relevant (host-derived, never a model
    /// claim).
    pub rank: u32,
}

/// Deterministic discovery result with its digest.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DiscoveryResult {
    /// Candidates in canonical order after bounding.
    pub candidates: Vec<DiscoveryCandidate>,
    /// Digest over the ordered candidates (`DiscoveryResult` v1).
    pub digest: String,
}

/// Inputs for [`discover_repository`].
pub struct DiscoveryInput<'a> {
    /// Unordered candidates.
    pub unordered_candidates: &'a [DiscoveryCandidateInput],
    /// Maximum candidates after bounding.
    pub max_candidates: usize,
    /// Normalized workspace-relative target paths of the task.
    pub task_targets: &'a [String],
}

const STRONG_EVIDENCE: [&str; 5] = [
    "exact task target",
    "direct import",
    "structural relationship",
    "explicit architecture ownership",
    "test mapping",
];

/// Deterministic baseline discovery: exact task targets rank first
/// (rank 0), strong-evidence/verified rank 1, everything else rank 2;
/// ties break by canonical path; bounds apply last. Shuffled input
/// order cannot change the result.
pub fn discover_repository(
    input: &DiscoveryInput<'_>,
) -> Result<DiscoveryResult, ArtifactIdentityError> {
    let ranked: Vec<DiscoveryCandidate> = input
        .unordered_candidates
        .iter()
        .map(|candidate| {
            let exact_target = input.task_targets.contains(&candidate.path);
            let strong_evidence = candidate.relevance
                == DiscoveryRelevance::Verified
                || candidate
                    .evidence
                    .iter()
                    .any(|item| STRONG_EVIDENCE.contains(&item.as_str()));
            let rank = if exact_target {
                0
            } else if strong_evidence {
                1
            } else {
                2
            };
            DiscoveryCandidate {
                path: candidate.path.clone(),
                relevance: candidate.relevance,
                evidence: candidate.evidence.clone(),
                rank,
            }
        })
        .collect();
    let ordered = stable_sort_by_key(&ranked, |candidate| {
        format!("{:04}:{}", candidate.rank, candidate.path)
    });
    let bounded: Vec<DiscoveryCandidate> =
        ordered.into_iter().take(input.max_candidates).collect();
    let candidate_values: Vec<CanonicalValue> = bounded
        .iter()
        .map(|candidate| {
            CanonicalValue::Object(BTreeMap::from([
                (
                    "path".to_owned(),
                    CanonicalValue::Str(candidate.path.clone()),
                ),
                (
                    "relevance".to_owned(),
                    CanonicalValue::Str(
                        candidate.relevance.as_str().to_owned(),
                    ),
                ),
                (
                    "rank".to_owned(),
                    CanonicalValue::U64(u64::from(candidate.rank)),
                ),
                (
                    "evidence".to_owned(),
                    CanonicalValue::Array(
                        candidate
                            .evidence
                            .iter()
                            .map(|item| CanonicalValue::Str(item.clone()))
                            .collect(),
                    ),
                ),
            ]))
        })
        .collect();
    let digest = compute_artifact_digest(
        "DiscoveryResult",
        1,
        &CanonicalValue::Object(BTreeMap::from([(
            "candidates".to_owned(),
            CanonicalValue::Array(candidate_values),
        )])),
    )?
    .value;
    Ok(DiscoveryResult { candidates: bounded, digest })
}

#[cfg(test)]
mod tests {
    use super::{
        DiscoveryCandidateInput, DiscoveryInput, DiscoveryRelevance,
        discover_repository,
    };
    use crate::determinism::{list_ownership, resolve_owner};

    fn candidate(
        path: &str,
        relevance: DiscoveryRelevance,
    ) -> DiscoveryCandidateInput {
        DiscoveryCandidateInput {
            path: path.to_owned(),
            relevance,
            evidence: Vec::new(),
        }
    }

    #[test]
    fn exact_targets_rank_first_and_ties_break_by_path() {
        let candidates = [
            candidate("res://z.tscn", DiscoveryRelevance::Verified),
            candidate("res://a.gd", DiscoveryRelevance::Candidate),
            candidate("res://player.gd", DiscoveryRelevance::Verified),
        ];
        let targets = vec!["res://player.gd".to_owned()];
        let result = discover_repository(&DiscoveryInput {
            unordered_candidates: &candidates,
            max_candidates: 10,
            task_targets: &targets,
        })
        .expect("discovers");
        // Exact target first (rank 0), then verified (rank 1), then
        // candidate (rank 2); ties break by path.
        assert_eq!(
            result
                .candidates
                .iter()
                .map(|c| c.path.as_str())
                .collect::<Vec<_>>(),
            vec!["res://player.gd", "res://z.tscn", "res://a.gd"]
        );
        assert_eq!(result.candidates[0].rank, 0);
        assert_eq!(result.digest.len(), 64);
    }

    #[test]
    fn bounds_apply_after_ordering() {
        let candidates = [
            candidate("res://a.gd", DiscoveryRelevance::Verified),
            candidate("res://b.gd", DiscoveryRelevance::Candidate),
        ];
        let result = discover_repository(&DiscoveryInput {
            unordered_candidates: &candidates,
            max_candidates: 1,
            task_targets: &[],
        })
        .expect("discovers");
        assert_eq!(result.candidates.len(), 1);
        assert_eq!(result.candidates[0].path, "res://a.gd");
    }

    #[test]
    fn ownership_resolves_exact_and_overlap_aliases() {
        let exact = resolve_owner("tool projection").expect("exact match");
        assert_eq!(exact.owner, "ToolProjector");
        let overlap = resolve_owner("hashing").expect("overlap alias");
        assert_eq!(overlap.owner, "ArtifactDigest");
        assert!(resolve_owner("nonexistent responsibility").is_none());
        let listed = list_ownership();
        assert!(listed.len() >= 18);
        assert!(
            listed
                .windows(2)
                .all(|pair| pair[0].responsibility <= pair[1].responsibility)
        );
    }
}
